### PR TITLE
Fix bug in bfloat16 optimizer related to checkpointing

### PR DESCRIPTION
**Pull request by okoge-kaz**
## Overview

Fixed conditional branching where errors occurred.

```diff
-        elif self.bfloat16_enabled() and not self.zero_optimization():
+        elif self.bfloat16_enabled() and hasattr(self.optimizer, "bf16_groups"):
```

## Error Detail

I encountered the following error while training with zero 1 optimization and bf16 using Megatron-DeepSpeed.

```
.env/lib/python3.11/site-packages/deepspeed/runtime/engine.py", line 3288, in _get_zero_param_shapes
    ) == 2 else self.optimizer.fp16_groups
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
    ) == 2 else self.optimizer.fp16_groups
    ) == 2 else self.optimizer.fp16_groups
AttributeError: 'BF16_Optimizer' object has no attribute 'fp16_groups'
    ) == 2 else self.optimizer.fp16_groups
AttributeError: 'BF16_Optimizer' object has no attribute 'fp16_groups'. Did you mean: 'bf16_groups'?
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
                ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'BF16_Optimizer' object has no attr
```

and also when loading the checkpoint, I got another error.

```
.env/lib/python3.11/site-packages/deepspeed/runtime/engine.py", line 2873, in _load_zero_checkpoint
    self.optimizer.load_state_dict(state_dict_list=zero_sd_list,
TypeError: BF16_Optimizer.load_state_dict() got an unexpected keyword argument 'load_serial'
```

## Related Pull Request 

https://github.com/microsoft/DeepSpeed/pull/3759

## Related Issue

https://github.com/microsoft/DeepSpeed/issues/4272

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -2868,11 +2868,17 @@ def _load_zero_checkpoint(self, load_dir, tag, load_optimizer_states=True):
             if zero_sd_list is None:
                 return False
 
-        self.optimizer.load_state_dict(state_dict_list=zero_sd_list,
-                                       load_optimizer_states=load_optimizer_states,
-                                       load_from_fp32_weights=self.zero_load_from_fp32_weights(),
-                                       checkpoint_folder=checkpoint_folder,
-                                       load_serial=load_serial)
+        if self.zero_optimization_stage() == 3:
+            self.optimizer.load_state_dict(state_dict_list=zero_sd_list,
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:2872]: The preferred solution here is to add `load_serial` to bf16_optimizer's `load_state_dict()`. Our intention is to propagate that feature to all the optimizer implementations.

Comment (tjruwase) [reply deepspeed/runtime/engine.py:2872]: @ShijieZZZZ, FYI

**Review by tjruwase — APPROVED** [review]

**Comment by okoge-kaz** [comment]
I have confirmed that checkpoints can be saved & loaded with bf16 in Megatron-DeepSpeed!

(A100 40GB x 16(2node))

image

**Comment by tjruwase** [comment]
@okoge-kaz, thanks for sharing your results.

**Comment by tjruwase** [comment]
Fix #3759

**Comment by okoge-kaz** [comment]
@tjruwase All test passed! 
May I ask for a merge?

**Pull request merged**